oak_channel/SPEC.md

```diff
@@ -138,149 +93,80 @@ A byte encoded frame MUST consist of the following fields:
+be sent in sequential order of the frame number over the channel.
 
-```text
- 0                   1                   2                   3
- 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                         frame padding                         |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|          frame length         |          frame flags          |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|             frame body [ request_message length ]             |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|          frame body [ request_message invocation_id ]         |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|            frame body [ request_message method_id ]           |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|             frame body [ request_message padding ]            |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                                                               |
-+                                                               +
-|                                                               |
-+   frame body [ request_message body [ method parameters ] ]   +
-|                                                               |
-+                                                               +
-|                                                               |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+ ...
-```
-
-<!-- Diagram generated with https://www.luismg.com/protocol/, using the schema
-"frame padding:32,frame length:16,frame flags:16,frame body [ request_message length ]:32,frame body [ request_message invocation_id ]:32,frame body [ request_message method_id ]:32,frame body [ request_message padding ]:32,frame body [ request_message body [ method parameters ] ]:128" -->
-
-#### Frame 1
+### Receiving Messages
 
-```text
- 0                   1                   2                   3
- 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                         frame padding                         |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|          frame length         |          frame flags          |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                                                               |
-+   frame body [ request_message body [ method parameters ] ]   +
-|                                                               |
-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-```
+To receive a message the recipent incrementally reads frames from the
+communication channel. Messages are parsed from frames by appending the frame
+bodies in the order they are received. The recipient SHOULD validate the
```

Comment:
Why SHOULD the recipient validate the sequence?

Reply:
+1 I would expect this to be MUST, and in fact the client should abort if an out-of-order frame is received.

Reply:
I disagree. I interpret MUST as something that will always break the protocol if it is ignored. Not validating this will not detect corruption if it happened, but it will not actively break correctly formatted things.

Reply:
`MUST   This word, or the terms "REQUIRED" or "SHALL", mean that the
   definition is an absolute requirement of the specification.`

`SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.`

(From https://www.rfc-editor.org/rfc/rfc2119)

I think this should be an absolute requirement of the specification, so MUST is appropriate.

Reply:
I will change it to MUST, but I think SHOULD could also be appropriate. In the case where the communcations channel is infallible and the sender is trusted to accurately follow the protocol it could be a valid decision to skip this validation.